## Retail Store Sales Analysis with Python & SQL

##### **Project Objective**
This project demonstrates end-to-end data analytics workflow using SQL and Python

The dataset contains retail transaction records with missing and inconsistent values

**Goals:**
1. Load data from CSV into SQL safely
2. Handle missing values correctly
3. Perform exploratory data analysis (EDA)
4. Visualize business insights

---

### SQL and Python Integration

Install SQLAlchemy and pymysql

*%pip install SQLAlchemy*

*%pip install pymysql*

---

### Import Required Libraries

In [2]:
# Pandas for data manipulation
import pandas as pd

# numpy for mathematical operations
import numpy as np

# SQLAlchemy for SQL-Python connection
from sqlalchemy import create_engine

# for Visualization
import matplotlib.pyplot as plt

# Preprocessing tools for processing data to train model
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

# Machine Learning for predictive imputation
from sklearn.ensemble import RandomForestRegressor

# To Check Model Accuracy
from sklearn.metrics import mean_absolute_error, mean_squared_error
from sklearn.model_selection import train_test_split


### Connect Python to MySQL

In [3]:
# Create a connection engine to MySQL
# Replace username, password, and database name as per your setup

engine = create_engine(
    "mysql+pymysql://root:Riy%401713@localhost/retail_db"
)

 *Explanation*

This engine allows Python to:

*Read data from SQL tables*

*Write processed data back to SQL if needed*

---

### Load SQL Table into Pandas DataFrame

In [4]:
# Read retail_sales table from SQL into pandas DataFrame
df = pd.read_sql("SELECT * FROM retail_sales", engine)

# Display first 10 rows
df.head(10)

,transaction_id,customer_id,category,item,price_per_unit,quantity,total_spent,payment_method,location,transaction_date,discount_applied
0,TXN_6867343,CUST_09,Patisserie,Item_10_PAT,18.5,10.0,185.0,Digital Wallet,Online,2024-04-08,True\r
1,TXN_3731986,CUST_22,Milk Products,Item_17_MILK,29.0,9.0,261.0,Digital Wallet,Online,2023-07-23,True\r
2,TXN_9303719,CUST_02,Butchers,Item_12_BUT,21.5,2.0,43.0,Credit Card,Online,2022-10-05,False\r
3,TXN_9458126,CUST_06,Beverages,Item_16_BEV,27.5,9.0,247.5,Credit Card,Online,2022-05-07,\r
4,TXN_4575373,CUST_05,Food,Item_6_FOOD,12.5,7.0,87.5,Digital Wallet,Online,2022-10-02,False\r
5,TXN_7482416,CUST_09,Patisserie,None,NaN,10.0,200.0,Credit Card,Online,2023-11-30,\r
6,TXN_3652209,CUST_07,Food,Item_1_FOOD,5.0,8.0,40.0,Credit Card,In-store,2023-06-10,True\r
7,TXN_1372952,CUST_21,Furniture,None,33.5,NaN,NaN,Digital Wallet,In-store,2024-04-02,True\r
8,TXN_9728486,CUST_23,Furniture,Item_16_FUR,27.5,1.0,27.5,Credit Card,In-store,2023-04-26,False\r
9,TXN_2722661,CUST_25,Butchers,Item_22_BUT,36.5,3.0,109.5,Cash,Online,2024-03-14,False\r


### Initial Data Understanding

In [5]:
# Check Dataset Shape

# Number of rows and columns
df.shape

(12575, 11)

In [6]:
# Check Column Data Types 

# Understand data types and null counts
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   transaction_id    12575 non-null  object 
 1   customer_id       12575 non-null  object 
 2   category          12575 non-null  object 
 3   item              11362 non-null  object 
 4   price_per_unit    11966 non-null  float64
 5   quantity          11971 non-null  float64
 6   total_spent       11971 non-null  float64
 7   payment_method    12575 non-null  object 
 8   location          12575 non-null  object 
 9   transaction_date  12575 non-null  object 
 10  discount_applied  12575 non-null  object 
dtypes: float64(3), object(8)
memory usage: 1.1+ MB


---

In [7]:
# Statistical summary of numeric columns
df.describe()

,price_per_unit,quantity,total_spent
count,11966.000000,11971.000000,11971.000000
mean,23.365912,5.536380,129.652577
std,10.743519,2.857883,94.750697
min,5.000000,1.000000,5.000000
25%,14.000000,3.000000,51.000000
50%,23.000000,6.000000,108.500000
75%,33.500000,8.000000,192.000000
max,41.000000,10.000000,410.000000


### Data Validation

In [8]:
# check duplicacy in the data 
df.duplicated().sum()

0

In [9]:
# check number of duplicate data of particular column
print("Number of Duplicate Transaction IDs:", df['transaction_id'].duplicated().sum())


Number of Duplicate Transaction IDs: 0


In [10]:
# Verify categorical consistency (e.g., checking for typos in 'category')
print("Unique categories found:")
print(df['category'].unique())

Unique categories found:
['Patisserie' 'Milk Products' 'Butchers' 'Beverages' 'Food' 'Furniture'
 'Electric household essentials' 'Computers and electric accessories']


### Converting Columns Data Types

#### Convert Date Column to Datetime

In [11]:
# Convert transaction_date to datetime format

df['transaction_date'] = pd.to_datetime(df['transaction_date'])
df['transaction_date'].dtypes

dtype('<M8[ns]')

#### Convert String Column to Boolean data type

In [12]:
df['discount_applied'].unique()

array(['True\r', 'False\r', '\r'], dtype=object)

In [13]:
# Remove extra whitespace and carriage return characters (\r)
# This cleans values like 'True\r', 'False\r', and '\r'
df['discount_applied'] = df['discount_applied'].astype(str).str.strip()
df['discount_applied'].unique()

array(['True', 'False', ''], dtype=object)

In [14]:
# Map cleaned string values to actual boolean values
# Any value not listed becomes NaN
df['discount_applied'] = df['discount_applied'].map({
   'True' : True,
   'False' : False 
})

In [15]:
# To check unique values and data type of the column
df['discount_applied'].unique()

array([True, False, nan], dtype=object)

In [16]:
# Convert discount applied data type to boolean datatype
df['discount_applied'] = df['discount_applied'].astype('boolean')
df['discount_applied'].dtype 

BooleanDtype

#### Convert Float data type to Integer

In [17]:
# Convert quantiy columnn data type 
df['quantity'] = df['quantity'].astype('Int64')
df['quantity'].unique()

<IntegerArray>
[10, 9, 2, 7, 8, <NA>, 1, 3, 6, 4, 5]
Length: 11, dtype: Int64

In [18]:
df.isna().sum()

transaction_id         0
customer_id            0
category               0
item                1213
price_per_unit       609
quantity             604
total_spent          604
payment_method         0
location               0
transaction_date       0
discount_applied    4199
dtype: int64

### Handling Missing Data
The dataset contains missing values in following columns:
- item
- price_per_unit
- quantity
- total_spent
- discount_applied

The Goal is to use valid business logic to fill missing values of these columns

#### Printing null values rate

In [19]:
for i in df.columns:
    null_rate = df[i].isnull().sum() / len(df)*100
    if null_rate>0:
        print(f"{i} null rate: {null_rate.round(2)}")

item null rate: 9.65
price_per_unit null rate: 4.84
quantity null rate: 4.8
total_spent null rate: 4.8
discount_applied null rate: 33.39


#### Fill item column 
To fill null values in item column, We will use Price Per Unit of the items 

As we know, each item number have its own PPU

- we will create a dictionary with PPU and Category as Key

- Use the dictionary to fill the Item column

In [20]:
# Create dictionary with (PPU, Category) as key
ppu_category_to_item = (
    df.dropna(subset=['item', 'price_per_unit', 'category'])
      .groupby(['price_per_unit', 'category'])['item']
      .first()
      .to_dict()
)

print("\nComposite mappings:\n")
print(ppu_category_to_item)


Composite mappings:

{(5.0, 'Beverages'): 'Item_1_BEV', (5.0, 'Butchers'): 'Item_1_BUT', (5.0, 'Computers and electric accessories'): 'Item_1_CEA', (5.0, 'Electric household essentials'): 'Item_1_EHE', (5.0, 'Food'): 'Item_1_FOOD', (5.0, 'Furniture'): 'Item_1_FUR', (5.0, 'Milk Products'): 'Item_1_MILK', (5.0, 'Patisserie'): 'Item_1_PAT', (6.5, 'Beverages'): 'Item_2_BEV', (6.5, 'Butchers'): 'Item_2_BUT', (6.5, 'Computers and electric accessories'): 'Item_2_CEA', (6.5, 'Electric household essentials'): 'Item_2_EHE', (6.5, 'Food'): 'Item_2_FOOD', (6.5, 'Furniture'): 'Item_2_FUR', (6.5, 'Milk Products'): 'Item_2_MILK', (6.5, 'Patisserie'): 'Item_2_PAT', (8.0, 'Beverages'): 'Item_3_BEV', (8.0, 'Butchers'): 'Item_3_BUT', (8.0, 'Computers and electric accessories'): 'Item_3_CEA', (8.0, 'Electric household essentials'): 'Item_3_EHE', (8.0, 'Food'): 'Item_3_FOOD', (8.0, 'Furniture'): 'Item_3_FUR', (8.0, 'Milk Products'): 'Item_3_MILK', (8.0, 'Patisserie'): 'Item_3_PAT', (9.5, 'Beverages'): 'It

##### Use the dictionary to fill the Item column

In [21]:
df['item'] = df.apply(
    lambda row :
    ppu_category_to_item.get((row['price_per_unit'], row['category']))
    if 
        pd.isna(row['item'])
        and not pd.isna(row['price_per_unit'])
        and not pd.isna(row['category'])
    else 
        row['item'],
    axis = 1
)

In [22]:
# Verify if any null values present in item column
df['item'].isnull().sum()

609

In [23]:
df['item'].unique()

array(['Item_10_PAT', 'Item_17_MILK', 'Item_12_BUT', 'Item_16_BEV',
       'Item_6_FOOD', None, 'Item_1_FOOD', 'Item_20_FUR', 'Item_16_FUR',
       'Item_22_BUT', 'Item_3_BUT', 'Item_2_FOOD', 'Item_24_PAT',
       'Item_16_MILK', 'Item_14_BEV', 'Item_17_PAT', 'Item_13_EHE',
       'Item_21_FUR', 'Item_7_BEV', 'Item_4_EHE', 'Item_10_FOOD',
       'Item_14_FUR', 'Item_24_FUR', 'Item_20_BUT', 'Item_25_FUR',
       'Item_14_FOOD', 'Item_22_PAT', 'Item_11_FOOD', 'Item_6_PAT',
       'Item_21_EHE', 'Item_13_PAT', 'Item_25_BEV', 'Item_23_FOOD',
       'Item_10_FUR', 'Item_11_BEV', 'Item_23_BUT', 'Item_22_BEV',
       'Item_10_EHE', 'Item_24_BUT', 'Item_8_BEV', 'Item_3_FOOD',
       'Item_12_FOOD', 'Item_16_CEA', 'Item_11_PAT', 'Item_16_BUT',
       'Item_5_CEA', 'Item_19_MILK', 'Item_23_FUR', 'Item_7_FUR',
       'Item_15_CEA', 'Item_6_MILK', 'Item_24_CEA', 'Item_22_CEA',
       'Item_22_FOOD', 'Item_2_BUT', 'Item_14_PAT', 'Item_12_PAT',
       'Item_18_FOOD', 'Item_1_PAT', 'Item_4_BEV', 'Ite

#### Fill Discount Applied column
Now the column have 33.39% null values and there is no possible way to find values,

Because the values are of boolean datatype and there's no pattern in the data

In this situation,
assuming that all null values means that the discount wasnt applied and we will return false for all empty values in the dataset.

In [24]:
# fill False where the value is null in discount column and convert column into boolean datatype
df['discount_applied'] = df['discount_applied'].fillna(False).astype(bool)

In [25]:
# Check if there is null value present in the column
df['discount_applied'].isnull().sum()

0

#### Fill price per unit column 
We have quantity and total spent column 

we can fill price per unit column by using simple mathematical formula, 

Price Per Unit = Total Spent / Quantity

In [26]:
df["price_per_unit"] = df.apply(
    lambda row:
    row["total_spent"] / row["quantity"]
    if 
        pd.isna(row["price_per_unit"]) 
        and not pd.isna(row["total_spent"]) 
        and not pd.isna(row["quantity"])
    else 
        row["price_per_unit"],
    axis=1
)

In [27]:
# Check if there is any null values in price column
df['price_per_unit'].isnull().sum()

0

#### Fill quantity and total spent column 

In [28]:
# Check if quantity column have the other values needed for calculation
print("Quantity null but have price and total spent values:")
df[df['quantity'].isna() & df['price_per_unit'].notna() & df['total_spent'].notna()].shape

Quantity null but have price and total spent values:


(0, 11)

In [29]:
# Check if total spent column have the other values needed for calculation
print("Total Spent null but have price and quantity values:")
df[df['total_spent'].isna() & df['price_per_unit'].notna() & df['quantity'].notna()].shape

Total Spent null but have price and quantity values:


(0, 11)

#### Observation

Since there is no possible way to fill the Quantity and Total Spent missing values 

Because both values are missing in the same row we cannot do mathematical operations on these columns

Therefore, we delete rows where there is null values in these two columns 

In [30]:
df = df.dropna()

### Final Validation after Cleaning Data

In [31]:
# Display clean data 
df.head(15)

,transaction_id,customer_id,category,item,price_per_unit,quantity,total_spent,payment_method,location,transaction_date,discount_applied
0,TXN_6867343,CUST_09,Patisserie,Item_10_PAT,18.5,10,185.0,Digital Wallet,Online,2024-04-08,True
1,TXN_3731986,CUST_22,Milk Products,Item_17_MILK,29.0,9,261.0,Digital Wallet,Online,2023-07-23,True
2,TXN_9303719,CUST_02,Butchers,Item_12_BUT,21.5,2,43.0,Credit Card,Online,2022-10-05,False
3,TXN_9458126,CUST_06,Beverages,Item_16_BEV,27.5,9,247.5,Credit Card,Online,2022-05-07,False
4,TXN_4575373,CUST_05,Food,Item_6_FOOD,12.5,7,87.5,Digital Wallet,Online,2022-10-02,False
6,TXN_3652209,CUST_07,Food,Item_1_FOOD,5.0,8,40.0,Credit Card,In-store,2023-06-10,True
8,TXN_9728486,CUST_23,Furniture,Item_16_FUR,27.5,1,27.5,Credit Card,In-store,2023-04-26,False
9,TXN_2722661,CUST_25,Butchers,Item_22_BUT,36.5,3,109.5,Cash,Online,2024-03-14,False
10,TXN_8776416,CUST_22,Butchers,Item_3_BUT,8.0,9,72.0,Cash,In-store,2024-12-14,True
12,TXN_5874772,CUST_23,Food,Item_2_FOOD,6.5,7,45.5,Cash,Online,2023-09-09,True


In [32]:
# checking number of rows and columns after deleting null values from these two columns
print("Number of rows and columns in the data:", df.shape)

# verifying the data type of all the columns
print("Datatype of all the columns:\n", df.dtypes)

Number of rows and columns in the data: (11362, 11)
Datatype of all the columns:
 transaction_id              object
customer_id                 object
category                    object
item                        object
price_per_unit             float64
quantity                     Int64
total_spent                float64
payment_method              object
location                    object
transaction_date    datetime64[ns]
discount_applied              bool
dtype: object


In [33]:
# verifying if still any null values are present in the dataset
df.isnull().sum()

transaction_id      0
customer_id         0
category            0
item                0
price_per_unit      0
quantity            0
total_spent         0
payment_method      0
location            0
transaction_date    0
discount_applied    0
dtype: int64

### Save the clean data in CSV file format

In [34]:
# make the copy of the clean dataset 
df_clean = df.copy()

In [35]:
# Save the cleaned data in your disk as csv file
df_clean.to_csv("retail_store_clean_data.csv", index=False)

In [36]:
# Save the cleaned data in your disk as excel file
df_clean.to_excel("retail_store_clean_data.xlsx", index=False)

### Final Conclusion
Data Cleaning & Preparation Summary

In this project, the retail transactions dataset was cleaned and prepared to ensure data accuracy, consistency, and readiness for business analysis.

The key data preparation steps included:

- The retail data was sourced directly from a SQL database and ingested into Python using Pandas

- Data quality assessment was performed to understand data structure, data types, missing values, and formatting issues prior to data cleaning

- Removal of formatting inconsistencies that could impact grouping and aggregation

- Converting columns into proper data type format to ensure accurate aggregations and calculations

- Handling missing values in columns such as discount_applied, price_per_unit, item, quanity and total spent using appropriate business-driven approaches

- Validation of data integrity after cleaning to ensure no critical fields remain incomplete

After validation, the cleaned dataset was saved as a separate file, preserving the original raw data and ensuring reproducibility for future analysis.

### Future Scope

- Perform exploratory data analysis to uncover sales patterns and seasonal trends
- Build interactive dashboards for management reporting
- Apply predictive models to forecast sales and customer demand